File: BaekJoon Online Judge/C++/Step12-Dynamic Programming_1/2156.cpp
//
//  2156.cpp
//  Step12-Dynamic Programming_1
//
//  Created by 김예빈 on 2019. 9. 10..
//  Copyright © 2019년 김예빈. All rights reserved.
//

#include <iostream>
using namespace std;

int main(int argc, const char * argv[]) {
    cin.tie(NULL);
    ios::sync_with_stdio(false);
    
    int n, i;
    
    cin >> n;
    
    int wine[n];
    int dp[n];
    
    for(i = 0; i < n; i++)
        cin >> wine[i];
    
    dp[0] = wine[0];
    dp[1] = wine[0] + wine[1] > wine[1] ? wine[0] + wine[1] : wine[1];
    dp[2] = wine[0] + wine[2] > wine[1] + wine[2] ? wine[0] + wine[2] : wine[1] + wine[2];
    
    dp[2] = dp[1] > dp[2] ? dp[1] : dp[2];
    
    for(i = 3; i < n; i++) {
        dp[i] = (dp[i - 3] + wine[i - 1] > dp[i - 2] ? dp[i - 3] + wine[i - 1] : dp[i - 2]) + wine[i];
        dp[i] = dp[i - 1] > dp[i] ? dp[i - 1] : dp[i];
    }

    cout << dp[n - 1];
    
    return 0;
}

File: BaekJoon Online Judge/Swift/Step22-Dynamic Programming_1/2156.swift
//
//  2156.swift
//  Step14-Dynamic Programming_1
//
//  Created by Yebin Kim on 2019/12/15.
//  Copyright © 2019 김예빈. All rights reserved.
//

import Foundation

if let N = Int(readLine()!) {
    var wine = Array<Int>(repeating: 0, count: 1)
    var dp = Array(repeating: 0, count: N + 1)
    
    for _ in 0..<N {
        wine.append(Int(readLine()!)!)
    }
    
    if N > 2 {
        dp[0] = wine[0]
        dp[1] = wine[0] + wine[1]
        dp[2] = max(wine[0] + wine[2], max(wine[1] + wine[2], dp[1]))
        
        for i in 3..<dp.count {
            dp[i] = max(dp[i - 2], wine[i - 1] + dp[i - 3])
            dp[i] += wine[i]
            dp[i] = max(dp[i - 1], dp[i])
        }
        
        print(dp.last!)
    } else {
        print(wine.reduce(0){ $0 + $1 })
    }
}
